File: SIHFinals/sih_backend/test_mongo_to_ml.js
// Test: MongoDB → ML API Integration
// This script fetches real user data from MongoDB and tests the ML model

import mongoose from 'mongoose';
import dotenv from 'dotenv';
import fetch from 'node-fetch';

dotenv.config();

// Youth Model Schema (simplified)
const youthSchema = new mongoose.Schema({
    mobile: String,
    fullname: String,
    email: String,
    qualification: String,
    skills: [String],
    experienceList: [{
        role: String,
        company: String,
        duration: String,
        description: String
    }]
});

const Youth = mongoose.model('Youth', youthSchema);

async function testMongoToML() {
    try {
        console.log('🔗 Connecting to MongoDB...');
        await mongoose.connect(process.env.MONGO_URI);
        console.log('✅ MongoDB Connected\n');

        // Fetch a user from database
        console.log('📊 Fetching user from database...');
        const users = await Youth.find().limit(5);

        if (users.length === 0) {
            console.log('❌ No users found in database!');
            console.log('💡 Create a user first by completing registration on frontend');
            process.exit(0);
        }

        console.log(`✅ Found ${users.length} users in database\n`);

        // Test with first user
        const user = users[0];
        console.log('👤 Testing with user:');
        console.log(`   Name: ${user.fullname || 'Not set'}`);
        console.log(`   Mobile: ${user.mobile}`);
        console.log(`   Email: ${user.email || 'Not set'}`);
        console.log(`   Qualification: ${user.qualification || 'Not set'}`);
        console.log(`   Skills: ${user.skills?.join(', ') || 'Not set'}`);
        console.log(`   Experience: ${user.experienceList?.length || 0} positions\n`);

        // Transform to ML format
        const mlPayload = {
            position: user.experienceList?.[0]?.role || "Entry Level",
            skills: user.skills?.join(", ") || "General",
            summary: user.experienceList?.map(exp => exp.description).join(" ") || "Seeking opportunities",
            qualification: user.qualification || "Bachelor's Degree",
            experience: user.experienceList?.length > 0 ? `${user.experienceList.length} positions` : "Fresher",
            work_experience: user.experienceList?.map(exp =>
                `${exp.role} at ${exp.company} for ${exp.duration}`
            ).join(". ") || "No prior experience"
        };

        console.log('🔄 Transformed data for ML API:');
        console.log(JSON.stringify(mlPayload, null, 2));
        console.log('');

        // Send to ML API
        console.log('🤖 Sending to ML API...');
        const ML_API_URL = process.env.ML_API_URL || 'http://localhost:5001';

        const response = await fetch(`${ML_API_URL}/match-jobs`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(mlPayload)
        });

        if (!response.ok) {
            throw new Error(`ML API returned status ${response.status}`);
        }

        const result = await response.json();

        console.log('✅ ML API Response Received!\n');
        console.log(`📊 Results:`);
        console.log(`   Total jobs searched: ${result.total_jobs_searched || 0}`);
        console.log(`   Matches found: ${result.matches?.length || 0}\n`);

        if (result.matches && result.matches.length > 0) {
            console.log('🎯 Top 5 Job Matches:\n');
            result.matches.slice(0, 5).forEach((job, index) => {
                console.log(`${index + 1}. ${job.title || 'N/A'}`);
                console.log(`   Company: ${job.company || 'N/A'}`);
                console.log(`   Location: ${job.location || 'N/A'}`);
                console.log(`   Match Score: ${(job.match_score * 100).toFixed(1)}%`);
                console.log(`   Skills Match: ${(job.breakdown?.skills_match * 100).toFixed(1)}%`);
                console.log(`   Experience Match: ${(job.breakdown?.experience_match * 100).toFixed(1)}%`);
                console.log('');
            });

            console.log('✅ SUCCESS! MongoDB → ML API integration is working!\n');
        } else {
            console.log('⚠️  No job matches found. ML API is working but no matches for this profile.\n');
        }

    } catch (error) {
        console.error('❌ Error:', error.message);
        console.error(error);
    } finally {
        await mongoose.connection.close();
        console.log('🔌 MongoDB connection closed');
    }
}

// Run the test
testMongoToML();

File: SIHFinals/test_mongo_to_ml.py
# Test: MongoDB → ML API Integration (Python Version)
# Fetches real user data from MongoDB and sends to ML model

from pymongo import MongoClient
import requests
import json
from dotenv import load_dotenv
import os

load_dotenv()

def test_mongo_to_ml():
    print("="*60)
    print("MongoDB → ML API Integration Test")
    print("="*60)
    print()
    
    # Connect to MongoDB
    print("🔗 Connecting to MongoDB...")
    MONGO_URI = os.getenv('MONGO_URI') or "mongodb+srv://sih_user:K6dJmnBUcBs8_Y_@cluster0.yb1popv.mongodb.net/sih_database?retryWrites=true&w=majority"
    
    try:
        client = MongoClient(MONGO_URI)
        db = client['sih_database']
        youths_collection = db['youths']
        
        print("✅ MongoDB Connected")
        print()
        
        # Fetch users
        print("📊 Fetching users from database...")
        users = list(youths_collection.find().limit(5))
        
        if not users:
            print("❌ No users found in database!")
            print("💡 Create a user first by:")
            print("   1. Go to http://localhost:3001")
            print("   2. Complete registration (mobile → OTP → password)")
            print("   3. Fill and submit youth form")
            print("   4. Then run this test again")
            return
        
        print(f"✅ Found {len(users)} user(s) in database")
        print()
        
        # Test with first user
        user = users[0]
        print("👤 Testing with user:")
        print(f"   Name: {user.get('fullname', 'Not set')}")
        print(f"   Mobile: {user.get('mobile', 'N/A')}")
        print(f"   Email: {user.get('email', 'Not set')}")
        print(f"   Qualification: {user.get('qualification', 'Not set')}")
        
        skills = user.get('skills', [])
        print(f"   Skills: {', '.join(skills) if skills else 'Not set'}")
        
        exp_list = user.get('experienceList', [])
        print(f"   Experience: {len(exp_list)} position(s)")
        print()
        
        # Transform to ML format
        ml_payload = {
            "position": exp_list[0].get('role', 'Entry Level') if exp_list else "Entry Level",
            "skills": ", ".join(skills) if skills else "General skills",
            "summary": " ".join([exp.get('description', '') for exp in exp_list]) if exp_list else "Seeking opportunities",
            "qualification": user.get('qualification', "Bachelor's Degree"),
            "experience": f"{len(exp_list)} positions" if exp_list else "Fresher",
            "work_experience": ". ".join([
                f"{exp.get('role', 'N/A')} at {exp.get('company', 'N/A')} for {exp.get('duration', 'N/A')}"
                for exp in exp_list
            ]) if exp_list else "No prior experience"
        }
        
        print("🔄 Transformed data for ML API:")
        print(json.dumps(ml_payload, indent=2))
        print()
        
        # Send to ML API
        print("🤖 Sending to ML API...")
        ML_API_URL = os.getenv('ML_API_URL') or 'http://localhost:5001'
        
        response = requests.post(
            f"{ML_API_URL}/match-jobs",
            json=ml_payload,
            headers={"Content-Type": "application/json"}
        )
        
        if response.status_code != 200:
            print(f"❌ ML API returned status {response.status_code}")
            print(response.text)
            return
        
        result = response.json()
        
        print("✅ ML API Response Received!")
        print()
        print("📊 Results:")
        print(f"   Total jobs searched: {result.get('total_jobs_searched', 0)}")
        print(f"   Matches found: {len(result.get('matches', []))}")
        print()
        
        matches = result.get('matches', [])
        if matches:
            print("🎯 Top 5 Job Matches:")
            print()
            for i, job in enumerate(matches[:5], 1):
                print(f"{i}. {job.get('title', 'N/A')}")
                print(f"   Company: {job.get('company', 'N/A')}")
                print(f"   Location: {job.get('location', 'N/A')}")
                print(f"   Match Score: {job.get('match_score', 0)*100:.1f}%")
                
                breakdown = job.get('breakdown', {})
                print(f"   Skills Match: {breakdown.get('skills_match', 0)*100:.1f}%")
                print(f"   Experience Match: {breakdown.get('experience_match', 0)*100:.1f}%")
                print()
            
            print("="*60)
            print("✅ SUCCESS! MongoDB → ML API integration is working!")
            print("="*60)
        else:
            print("⚠️  No job matches found.")
            print("   ML API is working but no matches for this profile.")
        
    except Exception as e:
        print(f"❌ Error: {e}")
        import traceback
        traceback.print_exc()
    
    finally:
        client.close()
        print()
        print("🔌 MongoDB connection closed")

if __name__ == "__main__":
    test_mongo_to_ml()
